### DISCUSSION TITLE
[rust] How to insert a Stmt during a fold?

### DISCUSSION THREAD

**Discussion by alshdavid** [swc-project/swc#8680] [Q&A]
I'm trying to insert a statement during a fold. I'm not sure what the correct approach for this is.

```rust
impl Fold for FoldStatements {
  fn fold_assign_expr(
    &mut self,
    assign :AssignExpr,
  ) -> AssignExpr {
    if my_condition(&assign) {
      // How do I do this?
      self.insert_after_stmt(Stmt::Expr(Expr::Lit(/* ... */)));
    }

    return assign;
  }
}
```

**Comment by kdy1** [comment]
You can store extra statements in the struct field like https://github.com/search?q=repo%3Aswc-project%2Fswc%20extra_stmts&type=code

https://github.com/swc-project/swc/blob/8938fb0ecf40b2bca322d896147e35e0985d27ce/crates/swc_ecma_transforms_proposal/src/decorator_2022_03.rs#L53

**Reply by alshdavid** [reply]
When do you add the statements?

For example, if analysing code like this:

```javascript
// React does this
module.exports.baz = function trigger() {
  module.exports.foo = 'bar'
}
```

I want to be able to transform it into 
```javascript
module[module_id].baz = function trigger() {
  module[module_id].foo = 'bar'
  module[module_id].default.foo = module[module_id].foo
}
module[module_id].default.baz = module[module_id].baz
```
If I add the statements to the struct and append them after the fold is complete
```rust
let mut fold_stmts = FoldStatements{ extra: vec![] };
let mut module = module.fold_with(&mut fold_stmts);
module.body.concat(fold_stmts.extra);
```
I'll end up with
```javascript
module[module_id].baz = function trigger() {
  module[module_id].foo = 'bar' // from fold_assign_expr
}
module[module_id].default.baz = module[module_id].baz
module[module_id].default.foo = module[module_id].foo // from "extra" vec
```

**Reply by alshdavid** [reply]
Come to think of it, that probably wouldn't work as expected as reassignments might cause issues. I should probably use getters for synthetic default exports - which can be added after the fold as you suggest

```javascript
module[module_id].baz = function trigger() {
  module[module_id].foo = 'bar' // from fold_assign_expr
}
Object.defineProperty(module[module_id].default, "baz", () => module[module_id].baz)
Object.defineProperty(module[module_id].default, "foo", () => module[module_id].foo)
```

**Reply by kdy1** [reply]
You can iterate over statements and inject statements after each iteration.

**Reply by alshdavid** [reply]
Hmmm, yeah I suppose. When factoring in all the block types and possible locations for an `AssignExpr`, that would involve quite a lot of match statements to recursively unpack the `Stmt`s to find the target `AssignExpr` to insert my statement after.

Feels a bit like writing my own `FoldWith` implementation at that point.

Is that the best way?
